### PR TITLE
Feature: Unbeatable Computer

### PR DESCRIPTION
**Pull request by bean00**
This PR also includes changes made for most recent refactoring (PR #8 Refactor to prep for unbeatable comp).

*I will not actually merge these branches. I'll just incorporate changes from your comments into a new branch off of master.

### CODE REVIEW

File: lib/computer_player.rb [hunk 1]
```diff
@@ -1,17 +1,80 @@
...
+  end
+
+  private
+
+  def minimax_score(board_string_array, next_player)
+    current_board = Board.new(@side_length)
+    current_board.set_board(board_string_array)
+    score_tracker = ScoreTracker.new(current_board)
+
+    player_that_just_went = (next_player == 1) ? 2 : 1
```
Comment (jdesrosiers) [inline lib/computer_player.rb:31]: Why do you use a number *and* a string to represent a player.  As far as I can tell, it's just extra work to maintain both and keep them in sync.  If you really need both, you should pull it into an abstraction and deal with just the abstraction.

File: lib/computer_player.rb [hunk 2]
```diff
@@ -1,17 +1,80 @@
...
+
+      index = scores.find_index(score)
+
+      @next_move = moves[index]
+      return score
+    end
+  end
+
+  def get_player_token(player)
+    if player == @player_num
```
Comment (jdesrosiers) [inline lib/computer_player.rb:73]: This looks like a good use case for the ternary operator

File: lib/computer_player.rb [hunk 3]
```diff
@@ -1,17 +1,80 @@
...
+
 class ComputerPlayer < Player
 
-  def initialize(player_num, board)
-    @board = board
+  def initialize(player_num, side_length)
+    @player_num = player_num
+    @next_move = 0
+    @side_length = side_length
+    @comp_token = "O"
```
Comment (jdesrosiers) [inline lib/computer_player.rb:10]: Symbols are a better choice for tokens.  They are much more efficient than strings.

File: lib/computer_player.rb [hunk 4]
```diff
@@ -1,17 +1,80 @@
...
   end
 
   def get_token
-    "O"
+    @comp_token
+  end
+
+  private
+
+  def minimax_score(board_string_array, next_player)
```
Comment (jdesrosiers) [inline lib/computer_player.rb:26]: Why pass a `board_string_array`?  Why not pass a `Board`?  Converting back and forth between `Board`s and `board_string_array`s seems like your way to clone a `Board`.  Wouldn't it be cleaner to ask the `Board` for a clone of itself?  Or, better yet, make your `Board` immutable.  Whenever the `Board` is altered, return a new `Board` with the new state.

File: lib/computer_player.rb [hunk 5]
```diff
@@ -1,17 +1,80 @@
...
+    current_board = Board.new(@side_length)
+    current_board.set_board(board_string_array)
+    score_tracker = ScoreTracker.new(current_board)
```
Comment (jdesrosiers) [inline lib/computer_player.rb:29]: It seems that the ScoreTracker object is created, used once, and then discarded.  This seems unnecessary.  If you pass the board into `get_player_score` instead, you can have one object that calculates scores instead of thousands objects that calculate scores.

File: lib/computer_player.rb [hunk 6]
```diff
@@ -1,17 +1,80 @@
...
+
+    current_score = score_tracker.get_player_score(@comp_token)
+
+    if score_tracker.is_game_finished?
+      return current_score
+    else
+      possible_moves = current_board.get_available_moves
+      
+      board_array = Array.new
+      possible_moves.length.times {
```
Comment (jdesrosiers) [inline lib/computer_player.rb:45]: Why do you do this loop twice (also line 54)?  Why store all these boards in an array before using them in the next loop?  The array doesn't seem to be used for anything.

File: lib/computer_player.rb [hunk 7]
```diff
@@ -1,17 +1,80 @@
...
+        board_array[i].move(move, next_player_token)
+        new_board_state = board_array[i].to_string_array
+
+        scores << minimax_score(new_board_state, player_that_just_went)
+        moves << move
+      }
+
+      score = (next_player == @player_num) ? scores.max : scores.min
+
+      index = scores.find_index(score)
```
Comment (jdesrosiers) [inline lib/computer_player.rb:65]: This is awkward.  It would be nicer if stored your results as an array of pairs instead of two separate arrays.

File: lib/human_player.rb [hunk 8]
```diff
@@ -2,20 +2,24 @@
...
 
-  def get_next_move
-    IOHandler.new.get_valid_move(@board.to_string_array)
+  def get_next_move(board)
+    IOHandler.new.get_valid_move(board.to_string_array)
   end
 
   def get_token
-    if @player_num == 1
+    if @playing_computer
```
Comment (jdesrosiers) [inline lib/human_player.rb:15]: Why does it matter if the opponent is the computer?  How is that relevant to which token the player is represented by?

Comment (jdesrosiers) [reply lib/human_player.rb:15]: This is a calculation on two fields that never change, but you recalculate the result each time `get_token` is called.  This should be calculated either in the constructor or externally and passed to the constructor.  `get_token` can be a simple accessor.

Comment (jdesrosiers) [reply lib/human_player.rb:15]: This expression is far too complicated.  There is no need for nested `if`s.

File: lib/score_tracker.rb [hunk 9]
```diff
@@ -4,27 +4,29 @@ class ScoreTracker
...
   end
 
   def has_either_player_won?
-    player_1_token = @player_tokens[0]
+    player_a_token = @player_tokens[0]
 
-    @player_scores[player_1_token] == WIN_SCORE ||
-    @player_scores[player_1_token] == LOSS_SCORE
+    @player_scores[player_a_token] == WIN_SCORE ||
+    @player_scores[player_a_token] == LOSS_SCORE
```
Comment (jdesrosiers) [inline lib/score_tracker.rb:29]: Indent here to make it clear these two lines are one expression.

File: lib/win_checker.rb [hunk 10]
```diff
@@ -18,9 +19,7 @@ def has_player_won?(token)
   private
 
   def get_full_sets(side_length)
```
Comment (jdesrosiers) [inline lib/win_checker.rb:21]: This method doesn't use the argument passed in.

**Review by jdesrosiers — CHANGES_REQUESTED** [review]

File: lib/computer_player.rb (hunk 2 above)
Comment (bean00) [reply lib/computer_player.rb:73]: get_player_token removed (not needed anymore)

File: lib/computer_player.rb (hunk 1 above)
Comment (bean00) [reply lib/computer_player.rb:31]: Not using the player number anymore. Instead, assume that X always goes first, and O goes second, and base turn on token (still want to leave option open for user to choose to go first or second against computer).

File: lib/human_player.rb (hunk 8 above)
Comment (bean00) [reply lib/human_player.rb:15]: Token is now passed in, and get_token is just an accessor.

File: lib/win_checker.rb (hunk 10 above)
Comment (bean00) [reply lib/win_checker.rb:21]: Removed the side_length argument.

File: lib/computer_player.rb (hunk 6 above)
Comment (bean00) [reply lib/computer_player.rb:45]: Only looping once now, and not using board_array anymore.

File: lib/computer_player.rb (hunk 4 above)
Comment (bean00) [reply lib/computer_player.rb:26]: Passing Board directly into minimax_score now.

File: lib/computer_player.rb (hunk 7 above)
Comment (bean00) [reply lib/computer_player.rb:65]: Using a hash, where the keys are moves and the values are scores (ex: { 1 => 0, 3 => -1, 4 => 1 })

File: lib/score_tracker.rb (hunk 9 above)
Comment (bean00) [reply lib/score_tracker.rb:29]: Indent added.

File: lib/computer_player.rb (hunk 5 above)
Comment (bean00) [reply lib/computer_player.rb:29]: Only creating one ScoreTracker object now. The board is being passed to update_scores, rather than get_player_score.

File: lib/computer_player.rb (hunk 3 above)
Comment (bean00) [reply lib/computer_player.rb:10]: For now I will still keep tokens as strings. I did try changing the tokens to symbols, to see what would happen, and I can get it to the point where there is 1 test failing (Controller play_game test for human vs. computer). I do plan on using symbols eventually.

**Comment by bean00** [comment]
All comments except for the tokens as symbols were addressed. Time for running tests has decreased from 5.3 sec to 1.6 sec.

**Comment by jdesrosiers** [comment]
You have a bunch of comments saying you made changes, but the code hasn't changed.  What's up?
